### PR TITLE
#58 : Implement sniff for class properties PHPDoc formatting

### PR DESCRIPTION
**Pull request by konarshankar07**
This PR added new rule for class properties for PHP Doc block
Fixed https://github.com/magento/magento-coding-standard/issues/58

### CODE REVIEW

**Comment by lenaorobei** [comment]
Looking at failed Travis and see that sniff produces false positive findings, for example it should not be triggered for method declaration.

**Comment by konarshankar07** [comment]
Hello @lenaorobei ,
I'm looking into this issue why its validating method declaration and I'll update you soon.
Thanks

**Comment by konarshankar07** [comment]
> @konarshankar07 please cover also cases with meaningless description. Examples:
> 
> ```
> /**
>      * @var Test
>      *
>      * Product Repository
>      */
>     private $productRepository;
> ```
> 
> Method for this check already exists https://github.com/magento/magento-coding-standard/blob/develop/Magento2/Sniffs/Commenting/PHPDocFormattingValidator.php#L58

Hello @lenaorobei ,
The mentioned method is used to check if the Doc block has any meaningful short description or not. As per you example, We need to check the formatting of PHP doc block. Correct me If I'm wrong.
Thanks

**Comment by lenaorobei** [comment]
@konarshankar07, yes. The idea behind this is that `$productRepository` variable already has meaningful name, so short description with `Product Repository` should raise a warning.

**Comment by konarshankar07** [comment]
Hello @lenaorobei ,
Can I get an update for this PR?
Thanks

**Comment by lenaorobei** [comment]
Hi @konarshankar07, sure. This sniff is not just about formatting. As per our conversation, please use `PHPDocFormattingValidator` to check the description and add these cases to the unit test.

File: Magento2/Tests/Commenting/ClassPropertyPHPDocFormattingUnitTest.inc [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+     * Short Description
+     */
+    private $variablehasAlreadyhavePrivateClassMember;
+
+    /**
+     * @var variablehasAlreadyhaveProtectedClassMember
+     *
+     * Short Description
+     */
+    protected $variablehasAlreadyhaveProtectedClassMember;
```
Comment (lenaorobei) [inline Magento2/Tests/Commenting/ClassPropertyPHPDocFormattingUnitTest.inc]: Need to add case:
```
/**
     * @var variablehasAlreadyhavePrivateClassMember
     *
     * Variable name
     */
    private $variableName;
```

File: Magento2/Tests/Commenting/ClassPropertyPHPDocFormattingUnitTest.inc [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+class Bar {
+
+    /**
+     * @var variablehasAlreadyhaveProtectedClassMember
+     *
+     * Short Description
+     */
+    public $variablehasAlreadyhavePublicClassMember;
```
Comment (lenaorobei) [inline Magento2/Tests/Commenting/ClassPropertyPHPDocFormattingUnitTest.inc]: Please follow `camelCase` naming convention.

**Comment by konarshankar07** [comment]
Hello @lenaorobei ,
Can I get any update here?
Thanks

**Comment by lenaorobei** [comment]
@konarshankar07 I will test it against Magento codebase and will get back to you.

File: Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    public function processMemberVar(File $phpcsFile, $stackPtr)
+    {
+        $tokens = $phpcsFile->getTokens();
+
+        $commentEnd = $phpcsFile->findPrevious($this->ignoreTokens, ($stackPtr - 1), null, true);
+        if ($commentEnd === false
+            || ($tokens[$commentEnd]['code'] !== T_DOC_COMMENT_CLOSE_TAG
+                && $tokens[$commentEnd]['code'] !== T_COMMENT)
+        ) {
+            $phpcsFile->addWarning('Missing class property doc comment', $stackPtr, 'Missing');
```
Comment (lenaorobei) [inline Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php]:
```suggestion
            $phpcsFile->addWarning('Missing PHP DocBlock for property', $stackPtr, 'Missing');
```

File: Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            return;
+        }
+        $commentStart = $tokens[$commentEnd]['comment_opener'];
+        $foundVar = null;
+        foreach ($tokens[$commentStart]['comment_tags'] as $tag) {
+            if ($tokens[$tag]['content'] === '@var') {
+                if ($foundVar !== null) {
+                    $error = 'Only one @var tag is allowed in a class property comment';
```
Comment (lenaorobei) [inline Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php]:
```suggestion
                    $error = 'Only one @var tag is allowed for property PHP DocBlock';
```

File: Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                    $phpcsFile->addWarning($error, $tag, 'DuplicateVar');
+                } else {
+                    $foundVar = $tag;
+                }
+            }
+        }
+
+        if ($foundVar === null) {
+            $error = 'Missing @var tag in class property comment';
```
Comment (lenaorobei) [inline Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php]:
```suggestion
            $error = 'Class properties must have type declaration using @var tag';
```

File: Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        if ($string === false || $tokens[$string]['line'] !== $tokens[$foundVar]['line']) {
+            $error = 'Content missing for @var tag in class property comment';
+            $phpcsFile->addWarning($error, $foundVar, 'EmptyVar');
+            return;
+        }
+
+        // Check if class has already have meaningful description
+        $isShortDescription = $phpcsFile->findPrevious(T_DOC_COMMENT_STRING, $commentEnd, $foundVar, false);
+        if ($this->PHPDocFormattingValidator->providesMeaning($isShortDescription, $commentStart, $tokens) !== true) {
+            $error = 'Variable member already have meaningful name';
```
Comment (lenaorobei) [inline Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php]:
```suggestion
            $error = 'Short description duplicates class property name.';
```

File: Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+        if ($foundVar === null) {
+            $error = 'Missing @var tag in class property comment';
+            $phpcsFile->addWarning($error, $stackPtr, 'MissingVar');
+            return;
+        }
+
+        $string = $phpcsFile->findNext(T_DOC_COMMENT_STRING, $foundVar, $commentEnd);
+        if ($string === false || $tokens[$string]['line'] !== $tokens[$foundVar]['line']) {
+            $error = 'Content missing for @var tag in class property comment';
```
Comment (lenaorobei) [inline Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php]:
```suggestion
            $error = 'Content missing for @var tag in class property declaration.';
```

**Comment by konarshankar07** [comment]
Thanks for feedback

**Comment by sivaschenko** [comment]
@konarshankar07 thank you for the pull request! Would you be able to resolve the test failures?

**Comment by konarshankar07** [comment]
Hello @sivaschenko 
Sure, I will work on the test failure 
Thanks

**Comment by konarshankar07** [comment]
@sivaschenko ready for review

File: Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        // Check if class has already have meaningful description after @var tag
+        $isShortDescriptionAfterVar = $phpcsFile->findNext(
+            T_DOC_COMMENT_STRING,
+            $foundVar + 4,
+            $commentEnd,
+            false,
+            null,
+            false
+        );
+        if ($this->PHPDocFormattingValidator->providesMeaning(
```
Comment (svera) [inline Magento2/Sniffs/Commenting/ClassPropertyPHPDocFormattingSniff.php:98]: This `if` block can be extracted into a separate method, as it basically repeats is the same as the check done for checking description before `@var` tag. This way, we reduce repetition and also function size, as `processMemeberSize()` is quite big, helping readability along the way

**Comment by svera** [comment]
@magento import pr to magento-commerce/magento-coding-standard

**Comment by magento-engcom-team** [comment]
@svera the pull request successfully imported.

**Pull request merged**